nfsv4/buffer.h

```diff
@@ -13,42 +13,37 @@ static inline u32 pad(u32 a, u32 to)
 }
 
 typedef struct buffer {
+    heap h;
     bytes start;
     bytes end;
-    bytes length;
+    bytes capacity;
     void *contents;
 } *buffer;
 
+// move
 typedef void *heap;
-
 #define allocate(__h, __b) (malloc(__b))
 #define deallocate(__h, __x, __len) (free(__x))
 
-
 static bytes length(buffer b) 
 {
     return b->end - b->start;
 }
-    
 
-static buffer allocate_buffer(heap h, bytes length){
+static buffer allocate_buffer(heap h, bytes capacity){
     buffer b = allocate(h, sizeof(struct buffer));
     memset(b, 0, sizeof(struct buffer));    
-    b->length = length + 100;
-    b->contents = allocate(h, b->length);
+    b->capacity = capacity;
+    b->contents = allocate(h, b->capacity);
     return b;
 }
 
-
-
 static inline void buffer_extend(buffer b, bytes len)
 {
-    // pad to pagesize
-
-    if (b->length < (b->end + len)) {
-        bytes oldlen = b->length;
-        b->length = 2*(oldlen+len);
-        void *new =  allocate(b->h, b->length);
+    if (b->capacity < (b->end + len)) {
+        bytes oldcap = b->capacity;
+        b->capacity = 2*(oldcap+len);
```

Comment:
I'd recommend new capacity as: `max(2 * oldcap, len)`. As is we can end up allocating too much space, e.g., if `len==oldcap+1` then we end up increasing by over 4x in one step. I compared another library, Java's `StringBuilder` and found that they use `max(2*oldcap+2, len)`.

Reply:
Ok, how about we just go with the classic algorithm for now and then do further tuning later in the context of priorities for performance optimization?